File: src/entity/MapDescriptor.java
package entity;

public class MapDescriptor {
    private static String binToHex(String binStr) {
        // Add padding to binary string
        int numPadBits = 8 - binStr.length() % 8;
        String binStrPadded = numPadBits == 8 ? binStr : binStr + "0".repeat(numPadBits);

        StringBuilder hexStr = new StringBuilder();

        for (int i = 0; i < binStrPadded.length(); i += 4) {
            String hexDigitBin = binStrPadded.substring(i, i + 4);
            int hexDigitValue = Integer.parseInt(hexDigitBin, 2);
            String hexDigit = Integer.toString(hexDigitValue, 16);
            hexStr.append(hexDigit);
        }

        return hexStr.toString().toUpperCase();
    }

    private static String hexToBin(String hexStr) {
        StringBuilder binStr = new StringBuilder();

        for (int i = 0; i < hexStr.length(); i++) {
            String hexDigit = Character.toString(hexStr.charAt(i));
            int hexDigitValue = Integer.parseInt(hexDigit, 16);
            String hexDigitBin = Integer.toString(hexDigitValue, 2);

            int numPadBits = 4 - hexDigitBin.length();
            binStr.append("0".repeat(numPadBits));
            binStr.append(hexDigitBin);
        }

        return binStr.toString();
    }

    public static Map generateMap(String[] strs) {
        String exploredBin = hexToBin(strs[0]);
        String obstaclesBin = hexToBin(strs[1]);

        int exploredCount = 2;
        int obstaclesCount = 0;

        Map map = new Map();

        for (int r = 0; r < Map.NUM_ROWS; r++) {
            for (int c = 0; c < Map.NUM_COLS; c++) {
                if (exploredBin.charAt(exploredCount) == '0') {
                    map.setCell(r, c, Cell.Unexplored);
                } else {
                    if (obstaclesBin.charAt(obstaclesCount) == '0')
                        map.setCell(r, c, Cell.Free);
                    else
                        map.setCell(r, c, Cell.Obstacle);

                    obstaclesCount++;
                }

                exploredCount++;
            }
        }

        return map;
    }

    public static String[] generateMapDescriptor(Map map) {
        String[] strs = new String[2];

        StringBuilder exploredBin = new StringBuilder();
        StringBuilder obstaclesBin = new StringBuilder();

        exploredBin.append("11");

        for (int r = 0; r < Map.NUM_ROWS; r++) {
            for (int c = 0; c < Map.NUM_COLS; c++) {
                Cell cell = map.getCell(r, c);
                switch (cell) {
                    case Unexplored:
                        exploredBin.append("0");
                        break;

                    case Free:
                        exploredBin.append("1");
                        obstaclesBin.append("0");
                        break;

                    case Obstacle:
                        exploredBin.append("1");
                        obstaclesBin.append("1");
                        break;

                    default:
                        break;
                }
            }
        }

        exploredBin.append("11");

        strs[0] = binToHex(exploredBin.toString());
        strs[1] = binToHex(obstaclesBin.toString());

        return strs;
    }
}

File: python/map_descriptor.py
import math
from constants import NUM_ROWS, NUM_COLS
from enums import Cell
from utils import print_map


def hex_to_bin(hex_str):
	bin_str = "{:b}".format(int(hex_str, 16))
	num_pad_bits = len(hex_str) * 4 - len(bin_str)
	return "0" * num_pad_bits + bin_str


def bin_to_hex(bin_str):
	hex_str = f"{int(bin_str, 2):X}"
	num_pad_bits = math.ceil(len(bin_str) / 4) - len(hex_str)
	return "0" * num_pad_bits + hex_str


def generate_map(explored_str, obstacle_str):
	map_grid = []

	explored_bin = hex_to_bin(explored_str)
	obstacle_bin = hex_to_bin(obstacle_str)

	explored_count = 2
	obstacle_count = 0

	for r in range(NUM_ROWS):
		row = []

		for c in range(NUM_COLS):
			is_explored = explored_bin[explored_count] == "1"
			explored_count += 1

			if is_explored:
				is_obstacle = obstacle_bin[obstacle_count] == "1"
				row.append(Cell.OBSTACLE if is_obstacle else Cell.FREE)
				obstacle_count += 1

			else:
				row.append(Cell.UNEXPLORED)

		map_grid.append(row)

	return map_grid


def generate_map_descriptor(map_grid):
	explored_bin = "11"
	obstacle_bin = ""

	for r in range(NUM_ROWS):
		for c in range(NUM_COLS):
			cell = map_grid[r][c]

			if cell == Cell.UNEXPLORED:
				explored_bin += "0"

			elif cell == Cell.FREE:
				explored_bin += "1"
				obstacle_bin += "0"

			elif cell == Cell.OBSTACLE:
				explored_bin += "1"
				obstacle_bin += "1"

	explored_bin += "11"

	if len(obstacle_bin) % 8 != 0:
		num_pad_bits = 8 - len(obstacle_bin) % 8
		obstacle_bin += "0" * num_pad_bits

	explored_str = bin_to_hex(explored_bin)
	obstacle_str = bin_to_hex(obstacle_bin)

	return explored_str, obstacle_str


def main():
	with open("maps/map1.txt", "r") as f:
		strs = f.read().split("\n")

	map_grid = generate_map(*strs)
	print(generate_map_descriptor(map_grid))
	print_map(map_grid)


if __name__ == "__main__":
	main()
